File: SOURCE_DOCSUITE/VecompSoftware.DocSuiteWeb.Presentation/ExportHelper.vb
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports Telerik.Web.UI
Imports VecompSoftware.Helpers
Imports System.Linq
Imports VecompSoftware.Helpers.ExtensionMethods

Public Class ExportHelper
    Private Shared Sub ExportToContentType(ByRef control As Control, ByRef page As Page, ByVal fileName As String, ByVal contentType As String)
        page.Response.Clear()
        page.Response.Buffer = True
        page.Response.ContentType = contentType
        page.Response.Charset = "ISO-8859-1" '"UTF-8"
        If fileName <> "" Then
            page.Response.AddHeader("content-disposition", String.Format("attachment;filename={0}", fileName))
        End If
        page.Response.Charset = ""
        page.EnableViewState = False
        Dim oStringWriter As New System.IO.StringWriter
        Dim oHtmlTextWriter As New System.Web.UI.HtmlTextWriter(oStringWriter)
        control.RenderControl(oHtmlTextWriter)
        page.Response.Write(oStringWriter.ToString())
        page.Response.End()
    End Sub

    Public Shared Sub ExportToExcel(ByRef control As Control, ByRef page As Page, Optional ByVal fileName As String = "")
        'export to excel
        Dim contentType As String = "application/msexcel"
        ExportToContentType(control, page, fileName, contentType)
    End Sub

    Public Shared Sub ExportToWord(ByRef control As Control, ByRef page As Page, Optional ByVal fileName As String = "")
        'export to excel
        Dim contentType As String = "application/msword"
        ExportToContentType(control, page, fileName, contentType)
    End Sub

    Public Shared Function TableFromGrid(gridColumns As GridColumnCollection, gridItems As GridDataItemCollection, Optional border As Boolean = True) As Table
        Dim table As DSTable = New DSTable()
        table.Table.Style.Add("border-collapse", "collapse")

        Dim index As Integer = -1
        Dim manageableColumns As IDictionary(Of Integer, GridColumn) = gridColumns.Cast(Of GridColumn) _
            .ToDictionary(Function(k)
                              index += 1
                              Return index
                          End Function, Function(k) k).Where(Function(c) c.Value.Visible AndAlso c.Value.Display AndAlso Not c.Value.UniqueName.Eq("ClientSelectColumn")).ToDictionary(Function(i) i.Key, Function(c) c.Value)
        Dim maxTableWidth As Double = manageableColumns.Sum(Function(c) c.Value.HeaderStyle.Width.Value)

        'Creo l'intestazione
        table.CreateEmptyRow()
        Dim cellHeaderStyle As DSTableCellStyle
        Dim cellPercentageWidth As Double = 0
        For Each column As GridColumn In manageableColumns.Values
            cellHeaderStyle = New DSTableCellStyle()
            cellPercentageWidth = (column.HeaderStyle.Width.Value / maxTableWidth) * 100
            cellHeaderStyle.Width = Unit.Percentage(cellPercentageWidth)
            cellHeaderStyle.Font.Bold = True
            cellHeaderStyle.LineBox = border

            table.CurrentRow.CreateEmpytCell()
            table.CurrentRow.CurrentCell.Text = column.HeaderText
            table.CurrentRow.CurrentCell.ApplyStyle(cellHeaderStyle)
        Next

        Dim cellItemStyle As DSTableCellStyle
        For Each gridDataItem As GridDataItem In gridItems
            table.CreateEmptyRow()
            For Each column As KeyValuePair(Of Integer, GridColumn) In manageableColumns
                If column.Key <= gridDataItem.Cells.Count Then
                    cellItemStyle = New DSTableCellStyle()
                    cellItemStyle.LineBox = border

                    table.CurrentRow.CreateEmpytCell()
                    ' Più offsetColumn per evitare le colonne inziali
                    table.CurrentRow.CurrentCell.Text = CellToString(column.Value, gridDataItem, gridDataItem.Cells(column.Key + 2))
                    table.CurrentRow.CurrentCell.ApplyStyle(cellItemStyle)
                End If
            Next
        Next

        Return table.Table
    End Function

    Private Shared Function CellToString(column As GridColumn, dataItem As GridDataItem, ByRef cell As TableCell) As String
        Dim retval As String = String.Empty

        Select Case column.ColumnType
            Case "CompositeTemplateExportableColumn"
                Dim exportableColumn As IExportableColumn = DirectCast(column, IExportableColumn)
                retval = exportableColumn.GetExportableValue(dataItem)
            Case Else
                If cell.Controls.Count > 1 Then
                    Select Case cell.Controls(1).GetType.ToString
                        Case "System.Web.UI.WebControls.Image", "System.Web.UI.WebControls.ImageButton"
                            retval = ImageToString(ReflectionHelper.GetProperty(cell.Controls(1), "ImageUrl").ToString())
                        Case "Telerik.Web.UI.RadButton"
                            Dim radButton As RadButton = DirectCast(cell.Controls(1), RadButton)
                            If Not String.IsNullOrEmpty(radButton.Image.ImageUrl) Then
                                retval = ImageToString(radButton.Image.ImageUrl)
                            End If
                        Case Else
                            retval = ReflectionHelper.GetProperty(cell.Controls(1), "Text").ToString()
                    End Select
                ElseIf cell.Controls IsNot Nothing _
                AndAlso cell.Controls.Count > 0 _
                AndAlso String.Compare(cell.Controls(0).GetType().ToString(), "System.Web.UI.DataBoundLiteralControl", True).Equals(0) Then
                    retval = DirectCast(cell.Controls.Item(0), DataBoundLiteralControl).Text
                Else
                    retval = cell.Text
                End If
        End Select

        Return retval
    End Function

    Private Shared Function ImageToString(ByVal path As String) As String
        Dim sRet As String = String.Empty
        Dim filename As String = System.IO.Path.GetFileNameWithoutExtension(path)

        Select Case filename.ToUpper
            Case "MAIL16_U", "MAIL32_U"
                sRet = "Uscita"
            Case "MAIL16_I", "MAIL32_I"
                sRet = "Ingresso"
            Case "MAIL16_IU", "MAIL32_IU"
                sRet = "Ingresso/Uscita"
            Case "DOCMANNULLA"
                sRet = "Pratica Annullata"
            Case "DOCMARCHIVIA"
                sRet = "Pratica Archiviata"
            Case "DOCMCHIUSURA"
                sRet = "Pratica Chiusura"
            Case "DOCMRIAPERTURA"
                sRet = "Pratica Riapertura"
            Case "ATTOANNULLATO"
                sRet = "Atto Annullato"
            Case "DELIBERAANNULLATA"
                sRet = "Delibera Annullata"
                'Cartella DocSuite
            Case "ATTO16", "ATTO32", "RESOLUTION16"
                sRet = "Atto"
            Case "COLLEGAMENTO16", "COLLEGAMENTO32"
                sRet = "Collegamento"
            Case "DELIBERA16", "DELIBERA32"
                sRet = "Delibera"
            Case "FASCICOLI16", "FASCICOLI32", "FASCICLE_OPEN"
                sRet = "Fascicoli"
            Case "FASCICOLOCLOSED16", "FASCICOLOCLOSED32"
                sRet = "Fascicolo Chiuso"
            Case "PRATICA16", "PRATICA32"
                sRet = "Pratica"
            Case "PRATICHE16", "PRATICHE32"
                sRet = "Pratiche"
                'Cartella File
            Case "ACROBAT16", "ACROBAT32"
                sRet = "PDF"
            Case "ALLEGATI16", "ALLEGATI32"
                sRet = "Allegati"
            Case "ATTI16", "ATTI32"
                sRet = "Atti"
            Case "CHECKOUT16", "CHECKOUT32"
                sRet = "CheckOut"
            Case "DIGITALSIGN16", "DIGITALSIGN32"
                sRet = "Firmato"
            Case "EXCEL16", "EXCEL32"
                sRet = "Excel"
            Case "EXPLORERHTM16", "EXPLORERHTM32"
                sRet = "HTML"
            Case "EXPLORERMHT16", "EXPLORERMHT32"
                sRet = "MHT"
            Case "FASCICOLO16", "FASCICOLO32"
                sRet = "Fascicolo"
            Case "IMAGE16", "IMAGE32"
                sRet = "Immagine"
            Case "MAIL16", "MAIL32"
                sRet = "Mail"
            Case "NONE16", "NONE32"
                sRet = String.Empty
            Case "NOTEPAD16", "NOTEPAD32"
                sRet = "Testo/ASCII"
            Case "POWERPOINT16", "POWERPOINT32"
                sRet = "PowerPoint"
            Case "PROTOCOLLO16", "PROTOCOLLO32"
                sRet = "Protocollo"
            Case "VERSION16", "VERSION32"
                sRet = "Versione"
            Case "WINWORD16", "WINWORD32"
                sRet = "Word"
            Case "XML16", "XML32"
                sRet = "XML"
            Case Else
                sRet = filename
        End Select

        Return sRet
    End Function

End Class

File: SOURCE_DOCSUITE/VecompSoftware.DocSuiteWeb.Gui.WebComponent/ExportHelper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Web.UI;
using System.Web.UI.WebControls;
using Telerik.Web.UI;
using VecompSoftware.Helpers;

namespace VecompSoftware.DocSuiteWeb.Gui.WebComponent
{

    public class ExportHelper
    {
        /// <summary>
        /// Esporta il contenuto presente nella tabella HTML.
        /// Viene utilizzato l'oggetto PAGE per creare la risposta
        /// </summary>
        /// <param name="control"></param>
        /// <param name="page"></param>
        /// <param name="fileName"></param>
        /// <param name="contentType"></param>
        private static void ExportToContentType(Control control, Page page, string fileName, string contentType)
        {
            page.Response.Clear();
            page.Response.Buffer = true;
            page.Response.ContentType = contentType;
            page.Response.ContentEncoding = Encoding.UTF8;
            page.Response.BinaryWrite(Encoding.UTF8.GetPreamble());
            //UTF-8
            if (!string.IsNullOrEmpty(fileName))
            {
                page.Response.AddHeader("content-disposition", string.Format("attachment;filename={0}", fileName));
            }
            page.Response.Charset = "UTF-8";
            page.EnableViewState = false;
            StringWriter oStringWriter = new StringWriter();
            HtmlTextWriter oHtmlTextWriter = new HtmlTextWriter(oStringWriter);
            control.RenderControl(oHtmlTextWriter);
            page.Response.Write(oStringWriter.ToString());
            page.Response.Flush();
            page.Response.SuppressContent = true;
        }

        /// <summary>
        /// Setto il content per i file EXCEL type e esporto i dati presenti nella tabella
        /// </summary>
        /// <param name="control"></param>
        /// <param name="page"></param>
        /// <param name="fileName"></param>
        public static void ExportToExcel(Control control, Page page, string fileName = "")
        {
            //export to excel
            string contentType = "application/msexcel";
            ExportToContentType(control, page, fileName, contentType);
        }

        /// <summary>
        /// Setto il content per i file WORD type e esporto i dati presenti nella tabella
        /// </summary>
        /// <param name="control"></param>
        /// <param name="page"></param>
        /// <param name="fileName"></param>
        public static void ExportToWord(Control control, Page page, string fileName = "")
        {
            //export to excel
            string contentType = "application/msword";
            ExportToContentType(control, page, fileName, contentType);
        }

        /// <summary>
        /// Creazione della tabella di esportazione a partire dalle colonne e dagli item presenti nella griglia.
        /// </summary>
        /// <param name="gridColumns"></param>
        /// <param name="gridItems"></param>
        /// <param name="border"></param>
        /// <returns></returns>
        public static Table TableFromGrid(GridColumnCollection gridColumns, GridDataItemCollection gridItems, bool border = true)
        {
            Table table = new Table();
            table.Style.Add("border-collapse", "collapse");

            List<KeyValuePair<int, List<string>>> columns = new List<KeyValuePair<int, List<string>>>();
            //Recupero estraggo le informazioni delle colonne presenti nella griglia
            double width = 0;
            int fanta = 0;
            foreach (GridColumn col in gridColumns)
            {
                if (col.Visible && col.Display && col.UniqueName != "ClientSelectColumn")
                {
                    List<string> list = new List<string>();
                    list.Add(col.HeaderText);
                    list.Add(col.ColumnType);
                    list.Add(col.HeaderStyle.Width.Value.ToString());
                    KeyValuePair<int, List<string>> column = new KeyValuePair<int, List<string>>(fanta, list);
                    columns.Add(column);
                    width += col.HeaderStyle.Width.Value;
                }
                fanta = fanta + 1;
            }
            
            //Creo l'intestazione
            TableRow rowHeader = new TableRow();
            foreach (KeyValuePair<int, List<string>> item in columns)
            {
                TableCell cellStyle = new TableCell();
                double wPerc = (Convert.ToDouble(item.Value[2]) / width) * 100;
                cellStyle.Width = Unit.Percentage(wPerc);
                cellStyle.Font.Bold = true;
                cellStyle.BorderStyle = BorderStyle.Solid;
                cellStyle.BorderWidth = Unit.Pixel(1);
                cellStyle.Text = item.Value[0];
                rowHeader.Cells.Add(cellStyle);                 
            }
            table.Rows.Add(rowHeader);   
            
            //Creo i righe contenenti i dati
            GridDataItemCollection items = gridItems;
            foreach (GridDataItem item in items)
            {
                TableRow rowData = new TableRow();
                foreach (KeyValuePair<int, List<string>> x in columns)
                {
                    if (x.Key >= 0 && x.Key <= item.Cells.Count)
                    {
                        TableCell cellStyle = new TableCell();
                        cellStyle.BorderStyle = BorderStyle.Solid;
                        cellStyle.BorderWidth = Unit.Pixel(1);
                        // Più 2 per evitare le due colonne iniziali --> calcolo al volo l'offset per eliminare le prime colonne.
                        int offset = Math.Abs(columns.Count - item.Cells.Count);
                        cellStyle.Text = CellToString(item.Cells[x.Key + offset]);
                        rowData.Cells.Add(cellStyle);
                    }
                }
                table.Rows.Add(rowData);
            }
            return table;           
        }

        /// <summary>
        /// Estrazione del testo dalla cella
        /// </summary>
        /// <param name="cell"></param>
        /// <returns></returns>
        private static string CellToString(TableCell cell)
        {   
            string retval = string.Empty;
            
            if (cell.Controls.Count > 1)
            {
                switch (cell.Controls[1].GetType().ToString())
                {
                    case "System.Web.UI.WebControls.Image":
                    case "System.Web.UI.WebControls.ImageButton":
                        retval = ImageToString(ReflectionHelper.GetProperty(cell.Controls[1], "ImageUrl").ToString());
                        break;
                    default:
                        retval = ReflectionHelper.GetProperty(cell.Controls[1], "Text").ToString();
                        break;
                }
            }
            else if (cell.Controls != null && cell.Controls.Count > 0 && string.Compare(cell.Controls[0].GetType().ToString(), "System.Web.UI.DataBoundLiteralControl", true).Equals(0))
            {
                retval = ((DataBoundLiteralControl)cell.Controls[0]).Text;
            }
            else
            {
                retval = cell.Text;
            }
            
            return retval;
        }

        /// <summary>
        /// Converte il nome dell'immagine in testo da inserire nella tabella di esportazione
        /// </summary>
        /// <param name="path"></param>
        /// <returns></returns>
        public static string ImageToString(string path)
        {
            string sRet = string.Empty;
            string filename = System.IO.Path.GetFileNameWithoutExtension(path);

            switch (filename.ToUpper())
            {
                case "MAIL16_U":
                case "MAIL32_U":
                    sRet = "Uscita";
                    break;
                case "MAIL16_I":
                case "MAIL32_I":
                    sRet = "Ingresso";
                    break;
                case "MAIL16_IU":
                case "MAIL32_IU":
                    sRet = "Ingresso/Uscita";
                    break;
                case "DOCMANNULLA":
                    sRet = "Pratica Annullata";
                    break;
                case "DOCMARCHIVIA":
                    sRet = "Pratica Archiviata";
                    break;
                case "DOCMCHIUSURA":
                    sRet = "Pratica Chiusura";
                    break;
                case "DOCMRIAPERTURA":
                    sRet = "Pratica Riapertura";
                    break;
                case "ATTOANNULLATO":
                    sRet = "Atto Annullato";
                    break;
                case "DELIBERAANNULLATA":
                    sRet = "Delibera Annullata";
                    break;
                //Cartella DocSuite
                case "ATTO16":
                case "ATTO32":
                case "RESOLUTION16":
                    sRet = "Atto";
                    break;
                case "COLLEGAMENTO16":
                case "COLLEGAMENTO32":
                    sRet = "Collegamento";
                    break;
                case "DELIBERA16":
                case "DELIBERA32":
                    sRet = "Delibera";
                    break;
                case "FASCICOLI16":
                case "FASCICOLI32":
                    sRet = "Fascicoli";
                    break;
                case "FASCICOLOCLOSED16":
                case "FASCICOLOCLOSED32":
                    sRet = "Fascicolo Chiuso";
                    break;
                case "PRATICA16":
                case "PRATICA32":
                    sRet = "Pratica";
                    break;
                case "PRATICHE16":
                case "PRATICHE32":
                    sRet = "Pratiche";
                    break;
                //Cartella File
                case "ACROBAT16":
                case "ACROBAT32":
                    sRet = "PDF";
                    break;
                case "ALLEGATI16":
                case "ALLEGATI32":
                    sRet = "Allegati";
                    break;
                case "ATTI16":
                case "ATTI32":
                    sRet = "Atti";
                    break;
                case "CHECKOUT16":
                case "CHECKOUT32":
                    sRet = "CheckOut";
                    break;
                case "DIGITALSIGN16":
                case "DIGITALSIGN32":
                    sRet = "Firmato";
                    break;
                case "EXCEL16":
                case "EXCEL32":
                    sRet = "Excel";
                    break;
                case "EXPLORERHTM16":
                case "EXPLORERHTM32":
                    sRet = "HTML";
                    break;
                case "EXPLORERMHT16":
                case "EXPLORERMHT32":
                    sRet = "MHT";
                    break;
                case "FASCICOLO16":
                case "FASCICOLO32":
                    sRet = "Fascicolo";
                    break;
                case "IMAGE16":
                case "IMAGE32":
                    sRet = "Immagine";
                    break;
                case "MAIL16":
                case "MAIL32":
                    sRet = "Mail";
                    break;
                case "NONE16":
                case "NONE32":
                    sRet = string.Empty;
                    break;
                case "NOTEPAD16":
                case "NOTEPAD32":
                    sRet = "Testo/ASCII";
                    break;
                case "POWERPOINT16":
                case "POWERPOINT32":
                    sRet = "PowerPoint";
                    break;
                case "PROTOCOLLO16":
                case "PROTOCOLLO32":
                    sRet = "Protocollo";
                    break;
                case "VERSION16":
                case "VERSION32":
                    sRet = "Versione";
                    break;
                case "WINWORD16":
                case "WINWORD32":
                    sRet = "Word";
                    break;
                case "XML16":
                case "XML32":
                    sRet = "XML";
                    break;
                case "FASCICLE_PROCEDURE":
                    sRet = "Fascicolo di procedimento";
                    break;
                default:
                    sRet = filename;
                    break;
            }

            return sRet;
        }
    }
}